### PR TITLE
Functionality to support emitting Kafka Events for tracking

### PR DESCRIPTION
**Pull request by yogi1324**
Dear DIL maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### Description
This request consists of including metrics/events as part of DIL. In this initial commit we'll be emitting events from different stages of job execution. These events can be helpful in identifying critical aspects such as network bottle necks, automatic integration points such as number of flows, formats, functionalities etc for reporting purposes. The metadata can also be used for troubleshooting purposes as well

We need a reporter/client to emit the events built. This PR includes Kafka Based event implementation which uses Open source li-kafka-clients. Users can create their own Reporter by implementing `EventReporter` Interface.

All the Events have a generic schema `CdiTrackinhEvent` where metadata holds off of the essential information in the form of a Map. The `name` column is used for filtering the type of event. 

To include a new event, building an event can be included inside the `EventHelper` class.

### Tests
- The functionality is tested with Kafka internal environment.
- Included a Simple Integration Test along with a sample pull file to reference essential properties to turn on metrics

To use the metrics use following properties 

```
ms.metrics.enabled=true
ms.kafka.events.topic.name=<YourTopicName pereferred name CdiTrackingEvent>
bootstrap.servers=kafka.test.company.com:<port>
value.serializer=io.confluent.kafka.serializers.KafkaAvroSerializer
kafka.schemaRegistry.class=org.apache.gobblin.kafka.schemareg.LiKafkaSchemaRegistry
kafka.schemaRegistry.url=http://schemaregistry.test.company.com:<port>/schemaRegistry/schemas
```
Include SSL properties as defined in Kafka Documentation if you want to connect to Kafka through secured connection

### Next Steps

- Support for Events using Kafka Rest Client
- Support to accept Kafka SSL properties with `ms.ssl`
- Create more events such as pagination, workunit retry, end-of-job event

### CODE REVIEW

File: gradle/scripts/dependencyDefinitions.gradle [hunk 1]
```diff
@@ -179,5 +179,8 @@ ext.externalDependency = [
     "gobblin-runtime": "org.apache.gobblin:gobblin-runtime:" + gobblinVersion,
     "gobblin-sql": "org.apache.gobblin:gobblin-sql:" + gobblinVersion,
     "gobblin-utility": "org.apache.gobblin:gobblin-utility:" + gobblinVersion,
-    "commonsValidator": "commons-validator:commons-validator:1.6"
+    "commonsValidator": "commons-validator:commons-validator:1.6",
+    "li-apache-kafka-clients" : "com.linkedin.kafka.clients:li-apache-kafka-clients:1.0.39",
```
Comment (chris9692) [inline gradle/scripts/dependencyDefinitions.gradle:183]: Can this compile outside of LinkedIn?

**Review by chris9692 — APPROVED** [review]
Hi Yogesh, thanks for the thorough PR, great details, very easy to understand! I had a few comments/suggestions, please check.

**Comment by chris9692** [comment]
Yogesh will address comments in a separate PR.

File: cdi-core/src/main/java/com/linkedin/cdi/event/EventHelper.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  public static CdiTrackingEvent createInitializationEvent(State state, String namespace) {
+    CdiTrackingEvent.Builder builder = CdiTrackingEvent.newBuilder();
+    builder.setName("SourceInitializationEvent")
+        .setNamespace(namespace)
+        .setTimestamp(System.currentTimeMillis())
+        .setMetadata(fillEssentialProperties(state));
+    return builder.build();
+  }
+
+  public static CdiTrackingEvent createWorkunitCreationEvent(State state, List<WorkUnit> workUnit, String namespace) {
```
[the remark this answers is not in the document]
Comment (yogi1324) [reply cdi-core/src/main/java/com/linkedin/cdi/event/EventHelper.java:31]: The intent here is to build the event based on job progress. This method is just intended to hold event building for workUnitCreation and won't be reused for other events.

**Pull request merged**